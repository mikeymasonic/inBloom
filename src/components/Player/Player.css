@import url('https://fonts.googleapis.com/css2?family=Kanit&display=swap');

.audioContainer {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;

  .playerContainer {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    width: 50vw;
    visibility: none;
    margin-bottom: 2em;
    border-radius: 5px;
    border: 1px outset #FBC9B1;
    box-shadow: 0px 0px 4px 4px #070E0D;
    background: linear-gradient(120deg, #d3b1fb, #d6b1fb, #e0b3fb, #edb5fb, #fbb7fa, #fbbaea, #fcbcdb, #fcbfce, #fcc1c3, #fcccc3, #fcd3c5, #fcd6c5);
  }

  .title {
    font-weight: bold;
    font-size: 0.9vw;
    align-self: center;
    margin: 0.2em;
    color: rgb(53,5,97);
    text-align: center;
    letter-spacing: 0.5vmax;
    width: 100%;
    font-family: 'Kanit', sans-serif;
    text-shadow: 1px 1px 0px rgba(7, 14, 13, 0.2);
  }

.Player {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 99%;
  margin: 0.2em;
  visibility: none;
  background: rgb(53,5,97);
  background: linear-gradient(94deg, rgba(53,5,97,1) 0%, rgba(32,3,58,1) 100%);
  color: #FBC9B1;
  overflow: hidden;

  .playerControls {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;

    button {
      display: flex;
      justify-content: center;
      align-items: center;
      width: 4.5vw;
      height: 6vh;
      background: none;
      border: none;
    }

    button:hover {
      transform: scale(1.1)
    }
  
    button:active {
      transform: scale(0.9);
    }
  }

.rhap_container {
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  line-height: 1;
  font-family: inherit;
  width: 100%;

  button {
    display: flex;
    justify-content: center;
    align-items: center;
    color: #FBC6B1;
    background-color: white;
    width: 4.5vw;
    height: 6vh;
    border-radius: 50%;
    font-size: 2.5em;
    background: none;
    border: none;
  }

  button:hover {
    transform: scale(1.1)
  }

  button:active {
    transform: scale(0.9);
  }

  div {
    display: flex;
    justify-content: center;
    align-items: center;
    color: #FBC9B1;
    user-select: none;
    -webkit-user-select: none;
    margin: 0 0.2em 0 0.2em;
    text-shadow: 1px 1px 1px rgba(7, 14, 13, 0.9);
  }

  }
}

.Playlist{
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  border-radius: 5px;
  height: 4.2em;
  width: 50vw;
  color: #FBC9B1;
  padding: 0.1em 0.2em 0 0.2em;
  margin-bottom: 1em;
  border: 1px outset #FBC9B1;
  box-shadow: 0px 0px 4px 4px #070E0D;
  background: linear-gradient(120deg, #d3b1fb, #d6b1fb, #e0b3fb, #edb5fb, #fbb7fa, #fbbaea, #fcbcdb, #fcbfce, #fcc1c3, #fcccc3, #fcd3c5, #fcd6c5);

  a {
    display: block;
    padding: 0.4em 0.1em 0 0.7em;
    width: 25%;
    height: 4vh;
    overflow: hidden;
    white-space: nowrap;  
    text-overflow: ellipsis;
    font-size: 0.8em;
    box-shadow: inset 0px 0px 1px 0px #FBC9B1;
    border-right: 1px inset #FBC9B1;
    background: rgb(53,5,97);
    background: linear-gradient(94deg, rgba(53,5,97,1) 0%, rgba(32,3,58,1) 100%);
    border-radius: 2px;
    text-shadow: 2px 2px 1px rgba(7, 14, 13, 1);
    opacity: 0.9;
  }

  a:hover {
    /* box-shadow: 0px 0px 2px 1px #FBC9B1; */
    opacity: 1;
  }

  a:active {
    color: #fcdccd;
    box-shadow: inset 0px 0px 1px 1px #FBC9B1;
  }

  .playlistTitle {
    font-weight: bold;
    font-size: 1.2vmax;
    align-self: center;
    margin: 0 0 0.2em 0;
    color: rgb(53,5,97);
    text-align: center;
    letter-spacing: 0.9vmax;
    width: 100%;
    font-family: 'Kanit', sans-serif;
    text-shadow: 1px 1px 0px rgba(7, 14, 13, 0.2);
  }

  .scroll {
    width: 100%;
    box-shadow: inset 0px 0px 2px 1px #FBC9B1;
    background-color: #0A0113;
  }

  .content {
    display: flex;
    flex-direction: row;
    justify-content: left;
    align-items: center;
  }

}

}

.instructions {
  width: 100%;
  text-align: center;
  margin: 0 0 2em 0;
  color: #FBC9B1;
  font-size: 1.2em;
  letter-spacing: 0.15em;
  -webkit-animation: fadein 5s; 
    -moz-animation: fadein 5s; 
    animation: fadein 5s;
}

@media (min-width: 1500px) {
  .audioContainer {

    .Playlist {
      height: 5.5em;

      a {
        font-size: 1em;
      }
    }
  }
}

@media (max-width: 1200px) {
  .audioContainer {

    .playerContainer {
      width: 55vw;

      .title {
        font-size: 1.2vw;
        letter-spacing: 0.7vmax;
      }
    }

    .Player {
      width: 99%;
    }

    .Playlist {
      width: 55vw;
    }

  }
}

@media (max-width: 667px) {
  .audioContainer {

    .playerContainer {
      width: 85vw;
      margin: 0;

      .title {
        font-size: 2.2vw;
        letter-spacing: 0.7vmax;
      }

      .Player {
        width: 99%;

        .playerControls {

          button {
            width: 8.5vw;
          }
        }

        .rhap_container {

          button {
            width: 9.5vw;
          }

          div {
            margin: 0 0.1em 0 0.1em;
          }
        }
      }
    }


    .Playlist {
      width: 85vw;
      height: 7vh;

      a {
        width: 35%;
      }
    }

  }
}

@keyframes fadein {
  from { opacity: 0; }
  to   { opacity: 1; }
}

/* Firefox < 16 */
@-moz-keyframes fadein {
  from { opacity: 0; }
  to   { opacity: 1; }
}

/* Safari, Chrome and Opera > 12.1 */
@-webkit-keyframes fadein {
  from { opacity: 0; }
  to   { opacity: 1; }
}
