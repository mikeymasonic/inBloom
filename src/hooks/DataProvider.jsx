import React, { createContext, useState, useContext } from 'react';
import PropTypes from 'prop-types';

const DataContext = createContext();

export const DataProvider = ({ children }) => {
  // const [song, setSong] = useState('../mp3/example2.mp3');
  const [song, setSong] = useState(null);
  const [mapLocation, setMapLocation] = useState(null);
  const [playerVisible, setPlayerVisible] = useState(false);
  const [visualizerDisplay, setVisualizerDisplay] = useState(false);

  const handleSong = (track) => {
    setSong(track);
    setMapLocation(track.mapLocation);
  };

  const handleMapLocation = (location) => {
    setMapLocation(location);
  };

  const handlePlayerVisible = (bool) => {
    setPlayerVisible(bool);
  };

  const handleVisualizerDisplay = (bool) => {
    setVisualizerDisplay(bool);
  };

  return (
    <DataContext.Provider
      value={{
        song,
        mapLocation,
        playerVisible,
        visualizerDisplay,
        handleSong,
        handleMapLocation,
        handlePlayerVisible,
        handleVisualizerDisplay,
      }}
    >
      {children}
    </DataContext.Provider>
  );
};

DataProvider.propTypes = {
  children: PropTypes.node,
};

// get state
export const useSong = () => {
  const { song } = useContext(DataContext);
  return song;
};

export const useMapLocation = () => {
  const { mapLocation } = useContext(DataContext);
  return mapLocation;
};

export const usePlayerVisible = () => {
  const { playerVisible } = useContext(DataContext);
  return playerVisible;
};

export const useVisualizerDisplay = () => {
  const { visualizerDisplay } = useContext(DataContext);
  return visualizerDisplay;
};

// get handlers
export const useHandleSong = () => {
  const { handleSong } = useContext(DataContext);
  return handleSong;
};

export const useHandleMapLocation = () => {
  const { handleMapLocation } = useContext(DataContext);
  return handleMapLocation;
};

export const useHandlePlayerVisible = () => {
  const { handlePlayerVisible } = useContext(DataContext);
  return handlePlayerVisible;
};

export const useHandleVisualizerDisplay = () => {
  const { handleVisualizerDisplay } = useContext(DataContext);
  return handleVisualizerDisplay;
};

