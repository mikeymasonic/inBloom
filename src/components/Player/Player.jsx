import React, { createRef } from 'react';
import AudioPlayer, { RHAP_UI } from 'react-h5-audio-player';
import { useSong, usePlayerVisible, useVisualizerDisplay, useHandleVisualizerDisplay } from '../../hooks/DataProvider';
import Playlist from '../Playlist/Playlist';
import Visualizer from '../Visualizer/Visualizer';
import styles from './Player.css';
import 'reactjs-popup/dist/index.css';
// import 'react-h5-audio-player/lib/styles.css';

const Player = () => {
  const song = useSong();
  const playerVisible = usePlayerVisible();
  const player = createRef();
  const visualizerDisplay = useVisualizerDisplay();
  const handleVisualizerDisplay = useHandleVisualizerDisplay();
  const playerTitle = `${song?.mapLocation} - ${song?.title}`.toUpperCase();

  const handleFullscreen = () => {
    handleVisualizerDisplay(!visualizerDisplay);
  };

  return (
    <>
      <section className={styles.audioContainer}>
        <Playlist style={{ visibility: playerVisible ? 'visible' : 'hidden', height: playerVisible ? 100 : 0 }} />

        <div className={styles.playerContainer} style={{ visibility: playerVisible ? 'visible' : 'hidden', height: playerVisible ? 100 : 0 }}>
          <section className={styles.title}>{playerTitle}</section>

          <section className={styles.Player} style={{ visibility: playerVisible ? 'visible' : 'collapse', height: playerVisible ? 100 : 0 }}>
            <section className={styles.playerControls}>
              <AudioPlayer
                ref={player}
                className={styles.rhap_container}
                src={song?.url}
                customProgressBarSection={
                  [
                    RHAP_UI.CURRENT_TIME,
                    <section key="slash">/</section>,
                    RHAP_UI.DURATION,
                  ]
                }
              />
              <button onClick={handleFullscreen}>Fullscreen</button>
            </section>
          </section>
        </div>

        {playerVisible && <Visualizer forwardRef={player}/>} 
        
        {!playerVisible && <p className={styles.instructions}>Select a garden to listen to:</p>} 
      </section>
    </>
  );
};

export default Player;
