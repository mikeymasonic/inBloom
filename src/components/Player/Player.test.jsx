import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { DataProvider } from '../../hooks/DataProvider';
import AudioPlayer from 'react-h5-audio-player';
import Player from './Player';

describe('Player component', () => {
  afterEach(() => cleanup());
  it('renders Player', () => {
    const { asFragment } = render(
      <Router>
        <DataProvider>
          <Player>
            <AudioPlayer />
          </Player>
        </DataProvider>
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
