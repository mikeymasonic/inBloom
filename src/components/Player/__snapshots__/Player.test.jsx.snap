// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player component renders Player 1`] = `
<DocumentFragment>
  <section
    class="Playlist"
    style="visibility: hidden; height: 0px;"
  >
    <div
      class="title"
    >
      Playlist:
    </div>
    <div
      class="scrollarea scroll"
    >
      <div
        class="scrollarea-content content"
        style="margin-top: 0px; margin-left: 0px;"
        tabindex="1"
      />
    </div>
  </section>
  <section
    class="Player"
    style="visibility: collapse; height: 0px;"
  >
    <section
      class="title"
    >
       - 
    </section>
    <div
      aria-label="Audio Player"
      class="rhap_container rhap_container"
      role="group"
      tabindex="0"
    >
      <audio
        preload="auto"
      />
      <div
        class="rhap_main rhap_stacked"
      >
        <div
          class="rhap_progress-section"
        >
          <div
            class="rhap_time rhap_current-time"
            id="rhap_current-time"
          >
            --:--
          </div>
          <section>
            /
          </section>
          <div
            class="rhap_time rhap_total-time"
          >
            --:--
          </div>
        </div>
        <div
          class="rhap_controls-section"
        >
          <div
            class="rhap_additional-controls"
          >
            <button
              aria-label="Disable Loop"
              class="rhap_button-clear rhap_repeat-button"
              type="button"
            >
              <svg
                focusable="false"
                height="1em"
                preserveAspectRatio="xMidYMid meet"
                style="transform: rotate(360deg);"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 5.27L3.28 4L20 20.72L18.73 22l-3-3H7v3l-4-4l4-4v3h6.73L7 10.27V11H5V8.27l-3-3M17 13h2v4.18l-2-2V13m0-8V2l4 4l-4 4V7H8.82l-2-2H17z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
          <div
            class="rhap_main-controls"
          >
            <button
              aria-label="Rewind"
              class="rhap_button-clear rhap_main-controls-button rhap_rewind-button"
              type="button"
            >
              <svg
                focusable="false"
                height="1em"
                preserveAspectRatio="xMidYMid meet"
                style="transform: rotate(360deg);"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.5 12l8.5 6V6m-9 12V6l-8.5 6l8.5 6z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <button
              aria-label="Play"
              class="rhap_button-clear rhap_main-controls-button rhap_play-pause-button"
              type="button"
            >
              <svg
                focusable="false"
                height="1em"
                preserveAspectRatio="xMidYMid meet"
                style="transform: rotate(360deg);"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10 16.5v-9l6 4.5M12 2A10 10 0 0 0 2 12a10 10 0 0 0 10 10a10 10 0 0 0 10-10A10 10 0 0 0 12 2z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <button
              aria-label="Forward"
              class="rhap_button-clear rhap_main-controls-button rhap_forward-button"
              type="button"
            >
              <svg
                focusable="false"
                height="1em"
                preserveAspectRatio="xMidYMid meet"
                style="transform: rotate(360deg);"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13 6v12l8.5-6M4 18l8.5-6L4 6v12z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
          <div
            class="rhap_volume-controls"
          >
            <div
              class="rhap_volume-container"
            >
              <button
                aria-label="Mute"
                class="rhap_button-clear rhap_volume-button"
                type="button"
              >
                <svg
                  focusable="false"
                  height="1em"
                  preserveAspectRatio="xMidYMid meet"
                  style="transform: rotate(360deg);"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M14 3.23v2.06c2.89.86 5 3.54 5 6.71s-2.11 5.84-5 6.7v2.07c4-.91 7-4.49 7-8.77c0-4.28-3-7.86-7-8.77M16.5 12c0-1.77-1-3.29-2.5-4.03V16c1.5-.71 2.5-2.24 2.5-4M3 9v6h4l5 5V4L7 9H3z"
                    fill="currentColor"
                  />
                </svg>
              </button>
              <div
                aria-label="volume Control"
                aria-valuemax="100"
                aria-valuemin="0"
                aria-valuenow="100"
                class="rhap_volume-bar-area"
                role="progressbar"
                tabindex="0"
              >
                <div
                  class="rhap_volume-bar"
                >
                  <div
                    class="rhap_volume-indicator"
                    style="left: 100.00%; transition-duration: 0s;"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
  <p>
    select a garden to listen to:
  </p>
</DocumentFragment>
`;
